app/js/ExceptionHandler.js

```diff
@@ -0,0 +1,97 @@
+const Metrics = require('metrics-sharelatex')
+const logger = require('logger-sharelatex')
+
+// TODO: domain has been deprecated for some time - do we need it and is there a better way?
+
+// eslint-disable-next-line node/no-deprecated-api
+const domain = require('domain')
+
+const TWO_MINUTES = 120 * 1000
+
+class ExceptionHandler {
+  constructor() {
+    this._appIsOk = true
+  }
+
+  beginShutdown() {
+    if (this._appIsOk) {
+      this._appIsOk = false
+
+      // hard-terminate this process if graceful shutdown fails
+      const killTimer = setTimeout(() => process.exit(1), TWO_MINUTES)
+
+      if (typeof killTimer.unref === 'function') {
+        killTimer.unref()
+      } // prevent timer from keeping process alive
+
+      this.server.close(function() {
+        logger.log('closed all connections')
+        Metrics.close()
+        if (typeof process.disconnect === 'function') {
+          process.disconnect()
+        }
+      })
+      logger.log('server will stop accepting connections')
+    }
+  }
+
+  addMiddleware(app) {
+    app.use(this.middleware.bind(this))
+  }
+
+  appIsOk() {
+    return this._appIsOk
+  }
+
+  setNotOk() {
+    this._appIsOk = false
+  }
+
+  middleware(req, res, next) {
+    const rescueLogger = require('logger-sharelatex')
```

Comment:
Why can't we use the regular `logger` here?

Reply:
Good question. Primarily I'm copying the pattern that was already here, in that the logger was previously also re-required in the error handler. This could well have been bad code in the first place, rather than on purpose, back in the original CoffeeScript.

My assumption was that as this is effectively a global exception handler, we might have stuffed up the logger object and would need a new one, but I'm not sure whether re-requiring it here would be that helpful, considering what happens with caching and whatnot.

@briangough originally made the change to include the new logger in https://github.com/overleaf/filestore/commit/6842c7dd0a42acdcafdb70bb6be40515192db2ff so this looks like it was done on purpose, but 2014 was a long time ago and perhaps the reason for this is no longer relevant. :wink:

Reply:
It was even part of the initial commit https://github.com/overleaf/filestore/commit/5881a3ea2c913d5e004199916d4baa99f29f2c1e#diff-66ac8d4b3a5e0f5ee69c82648d9ade91R39.

There is just one `logger-sharelatex` object, which will be pulled in via the second `require`. Calling `.initialize` resets parts of it - creating a new `Bunyan` logger instance and streams. But we are not calling `.initialize` here.

So I would opt for removal here.

Reply:
Yeah, I think you're right. I'd missed that it was there at the beginning. :+1: